r/R/dplyr-summarize.R

```diff
@@ -322,15 +301,76 @@ arrow_eval_or_stop <- function(expr, mask) {
   out
 }
 
+# This function returns a list of expressions that can be used to project the
+# data before an aggregation to only the fields required for the aggregation,
+# including the fields used in the aggregations (the "targets") and the group
+# fields. The names of the returned list are used to ensure that the projection
+# node is wired up correctly to the aggregation node.
 summarize_projection <- function(.data) {
   c(
-    map(.data$aggregations, ~ .$data),
+    unlist(unname(imap(
+      .data$aggregations,
+      ~set_names(
+        .x$data,
+        aggregate_target_names(.x$data, .y)
+      )
+    ))),
+    .data$selected_columns[.data$group_by_vars]
+  )
+}
+
+# This function determines what names to give to the fields used in aggregations
+# (the "targets"). When an aggregate function takes 2 or more fields as targets,
+# this function gives the fields unique names by appending `..1`, `..2`, etc.
+# When an aggregate function is nullary, this function returns a zero-length
+# character vector.
+aggregate_target_names <- function(data, name) {
+  if (length(data) > 1) {
+    paste(name, seq_along(data), sep = "..")
+  } else if (length(data) > 0) {
+    name
+  } else {
+    character(0)
+  }
+}
+
+# This function returns a list of expressions representing the aggregated fields
+# that will be returned by an aggregation
+aggregated_fields <- function(aggs) {
+  map(
+    aggs,
+    ~Expression$create(.$fun, args = .$data, options = .$options)
+  )
+}
+
+# Unlike with other pairs of non-hash/hash aggregate kernels in the Arrow C++
+# library, the `tdigest` and `hash_tdigest` kernels have different output types.
```

Comment:
Perhaps a safer way to do the type determination is to build expressions with the hash versions. The prose documentation says the hash kernels are unary, but according to the error messages you get, as well as the inline function docs in hash_aggregate.cc, they're binary. After some trial and error, I got this to work:

```
> x <- Expression$create("tdigest", Expression$field_ref("a"), options=list(q=c(.1, .2)))
> x$type(schema(a = int32()))
Float64
double

> x <- Expression$create("hash_tdigest", Expression$field_ref("a"), Expression$scalar(1L)$cast(uint32()), options=list(q=c(.1, .2)))
> x$type(schema(a = int32()))
FixedSizeListType
fixed_size_list<item: double>[2]
```

So the second argument is the group hash, and that is `uint32()`.

Reply:
Very cool! 

I changed it to work like this, but I used `Scalar$create(1L, uint32())` instead of `Expression$scalar(1L)$cast(uint32())` as the groups argument.